### PR TITLE
JIT through ragged axes

### PR DESCRIPTION
**Pull request by axch**

### CODE REVIEW

File: jax/_src/interpreters/pxla.py [hunk 1]
```diff
@@ -2127,7 +2131,8 @@ def _to_logical_sharding(
   elif isinstance(aval, core.AbstractToken):
     return None
   else:
-    raise TypeError(aval)
+    assert isinstance(sharding, sharding_impls.XLACompatibleSharding)
```
Comment (yashk2810) [inline jax/_src/interpreters/pxla.py:2134]: I would suggest creating another elif for dynamic aval instead of returning sharding by default.

Comment (axch) [reply jax/_src/interpreters/pxla.py]: Done.

**Review by mattjj — APPROVED** [review]
Awesome!

File: jax/_src/pjit.py [hunk 2]
```diff
@@ -1429,7 +1431,9 @@ def _pjit_batcher(insert_axis, spmd_axis_name,
     name=name,
     keep_unused=keep_unused,
     inline=inline)
-  return vals_out, axes_out
+  resovled_axes_out = batching.resolve_ragged_axes_against_inputs_outputs(
```
Comment (yashk2810) [inline jax/_src/pjit.py:1434]: I posted this on the CL too but typo?

**Pull request merged**